pep-0XXX-environment-markers.rst

```diff
@@ -0,0 +1,184 @@
+PEP: 0XXX
+Title: Environment Markers
+Version: $Revision$
+Last-Modified: $Date$
+Author: James Polley <jp@jamezpolley.com>
+Status: Draft
+Type: Informational
+Content-Type: text/x-rst
+Created: 03-Jul-2015
+Post-History: 03-Jul-2015
+Replaces: 0345 (part of)
+
+
+Abstract
+========
+
+An **environment marker** describes a condition about the current execution
+environment. They are used to indicate when certain dependencies are only
+required in particular environments, and to indicate supported platforms
+for distributions with additional constraints beyond the availability of a
+Python runtime.
+
+Environment markers were first specified in PEP-0345[1]. PEP-0426[2] (which
+would replace PEP-0345) proposed extensions to the markers. When
+2.7.10 was released, even these extensions became insufficient due to
+their reliance on simple lexical comparisons, and thus this PEP has
+been born.
+
+
+Rationale
+=========
+
+Many Python packages are written with portability in mind.
+
+For many packages this means they aim to support a wide range of
+Python releases. If they depend on libraries such as argparse or
+simplejson - which started as external libraries, but later got
+incorporated into core - specifying a single set of requirements is
+difficult, as the set of required packages differs depending on the
+version of Python in use.
+
+For other packages, this means broad cross-platform support. However,
+this may mean (for instance) that certain packages are required for
+installation on Windows, but not other platforms.
+
+Environment Markers attempt to provide more flexibility in a list of
+requirements by allowing the developer to give some information about
+the conditions under which a requirement should be installed.
+
+
+Examples
+========
+
+Here are some examples of such markers::
+
+   "sys_platform == 'win32'"
+   "platform_machine == 'i386'"
+   "python_version == '2.4' or python_version == '2.5'"
+   "python_version>='2.7' and python_version<'3.0'"
+   "'linux' in sys_platform"
+
+And here's an example of some conditional metadata for a distribution that
+requires PyWin32 both at runtime and buildtime when using Windows::
+
+    "name": "ComfyChair",
```

Comment:
I'm not entirely comfortable with solely using the requirements file format for this PEP. We already have two^Wthree implementations with slightly different syntaxes - the format pip reads in requirements.txt, the format pbr reads in setup.py, and the format setuptools accepts as kwargs.

Even assuming we ignore pbr as being its own extra thing unrelated to the PEP, I'd still prefer to either explicitly show both formats; or that we simply focus on the part that doesn't vary (ie, "sys_platform == 'win32'") and document elsewhere how to embed an environment marker into the various places they can be used.

I think I prefer the former approach - or at least, to focus the document on the invariant part and then maybe have a section at the bottom showing markers embedded in requirements.txt and setup.py (and pbr's docs can document how to put a marker into setup.cfg)

I note that https://pip.pypa.io/en/latest/reference/pip_install.html?highlight=markers#requirement-specifiers says "Don't use quotes in requirement files."  I've updated the requirements.txt examples here to not use quotes.

Given that setuptools doesn't actually support the syntax I'm documenting here, it sounds like it would be worth starting on an implementation of that as part of the implementation of this PEP? Once we have a working markerlib (or replacement) I imagine it won't be too difficult for setuptools to notice a marker in the requirements and pass it off to markerlib for evaluation (he says, with an abundance of optimism probably born of not having looked too closely at the code yet)